src/zjs_web_sockets.c

```diff
@@ -0,0 +1,867 @@
+// Copyright (c) 2017, Intel Corporation.
+
+#include <zephyr.h>
+#include <random.h>
+#include <sections.h>
+#include <errno.h>
+
+#include <net/nbuf.h>
+#include <net/net_if.h>
+#include <net/net_core.h>
+#include <net/net_context.h>
+#include <net/net_mgmt.h>
+#include "mbedtls/sha1.h"
+#include "mbedtls/base64.h"
+
+#if defined(CONFIG_NET_L2_BLUETOOTH)
+#include <bluetooth/bluetooth.h>
+#include <gatt/ipss.h>
+#include <bluetooth/conn.h>
+#endif
+
+// ZJS includes
+#include "zjs_util.h"
+#include "zjs_buffer.h"
+#include "zjs_callbacks.h"
+#include "zjs_event.h"
+#include "zjs_modules.h"
+#include "zjs_net_config.h"
+#include "zjs_zephyr_port.h"
+
+/*
+ * TODO:
+ *   - Implement WebSocket client side
+ */
+
+#define DEFAULT_WS_BUFFER_SIZE  256
+#define CHECK(x) \
+    ret = (x); if (ret < 0) { ERR_PRINT("Error in " #x ": %d\n", ret); return zjs_error(#x); }
+
+typedef enum {
+    UNCONNECTED = 0,
+    AWAITING_ACCEPT,
+    CONNECTED
+} ws_state;
+
+typedef enum {
+    WS_PACKET_CONTINUATION = 0x00,
+    WS_PACKET_TEXT_DATA = 0x01,
+    WS_PACKET_BINARY_DATA = 0x02,
+    WS_PACKET_PING = 0x09,
+    WS_PACKET_PONG = 0x0a
+} ws_packet_type;
+
+// server handle
+typedef struct server_handle {
+    struct net_context *tcp_sock;
+    jerry_value_t accept_handler;
+    jerry_value_t server;
+    bool track;
+    uint16_t max_payload;
+} server_handle_t;
+
+// WS connection handle, unique per server connection
+typedef struct ws_connection {
+    struct net_context *tcp_sock;
+    server_handle_t *server_handle;
+    uint8_t *rbuf;
+    uint8_t *wptr;
+    uint8_t *rptr;
+    struct ws_connection *next;
+    char *accept_key;
+    jerry_value_t server;
+    jerry_value_t conn;
+    zjs_callback_id trigger_msg_id;
+    zjs_callback_id accept_handler_id;
+    ws_state state;
+} ws_connection_t;
+
+// structure to hold a decoded WS packet
+typedef struct ws_packet {
+    uint8_t fin;
+    uint8_t rsv1;
+    uint8_t rsv2;
+    uint8_t rsv3;
+    uint8_t opcode;
+    uint8_t mask_bit;
+    uint32_t payload_len;
```

Comment:
Regarding mask, do you mean change it to a uint32? or can it remain a uint8 array? Its easier to do the unmasking if its an array.